File: examples/python/financial_report/ingest_data.py
import asyncio
import uuid

from semantic_retrieval.ingestion.data_sources.fs.file_system import FileSystem
from semantic_retrieval.document.metadata.in_memory_document_metadata_db import (
    InMemoryDocumentMetadataDB,
)

from semantic_retrieval.document_parsers.multi_document_parser import (
    MultiDocumentParser,
    ParserConfig,
)

from semantic_retrieval.access_control.always_allow_document_access_policy_factory import (
    AlwaysAllowDocumentAccessPolicyFactory,
)


from semantic_retrieval.transformation.document.text.separator_text_chunker import (
    SeparatorTextChunker,
)

from semantic_retrieval.data_store.vector_dbs.pinecone_vector_db import PineconeVectorDB

from semantic_retrieval.transformation.embeddings.openai_embeddings import OpenAIEmbeddings


async def main():
    # Create a new FileSystem instance
    fileSystem = FileSystem("examples/example_data/financial_report")

    # Load documents using the FileSystem instance
    rawDocuments = fileSystem.load_documents()

    # Initialize an in-memory metadata DB
    metadataDB = InMemoryDocumentMetadataDB()

    # Parse the raw documents
    parsedDocuments = await MultiDocumentParser().parse_documents(
        rawDocuments,
        ParserConfig(
            {
                "metadataDB": metadataDB,
                "accessControlPolicyFactory": AlwaysAllowDocumentAccessPolicyFactory(),
            }
        ),
    )

    # Initialize a document transformer
    documentTransformer = SeparatorTextChunker({"metadataDB": metadataDB})

    # Transform the parsed documents
    transformedDocuments = await documentTransformer.transform_documents(parsedDocuments)

    # Generate a new namespace using UUID
    namespace = str(uuid.uuid4())
    print(f"NAMESPACE: {namespace}")

    # Create a PineconeVectorDB instance and index the transformed documents
    pineconeVectorDB = PineconeVectorDB.from_documents(
        transformedDocuments,
        {
            "indexName": "test-financial-report-py",
            "namespace": namespace,
            "embeddings": OpenAIEmbeddings(),
            "metadataDB": metadataDB,
        },
    )

    # TODO: validate state of pineconeVectorDB


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/financial_report/ingest_data.ts
import { FileSystem } from "../../../src/ingestion/data-sources/fs/fileSystem";
import * as MultiDocumentParser from "../../../src/ingestion/document-parsers/multiDocumentParser";
import dotenv from "dotenv";
import { InMemoryDocumentMetadataDB } from "../../../src/document/metadata/inMemoryDocumentMetadataDB";
import { AlwaysAllowDocumentAccessPolicyFactory } from "../../../src/access-control/alwaysAllowDocumentAccessPolicyFactory";
import { SeparatorTextChunker } from "../../../src/transformation/document/text/separatorTextChunker";
import { PineconeVectorDB } from "../../../src/data-store/vector-DBs/pineconeVectorDB";
import { OpenAIEmbeddings } from "../../../src/transformation/embeddings/openAIEmbeddings";
import { v4 as uuid } from "uuid";

dotenv.config();

async function main() {
  const fileSystem = new FileSystem(
    "examples/example_data/financial_report/10ks"
  );
  const rawDocuments = await fileSystem.loadDocuments();
  const metadataDB = new InMemoryDocumentMetadataDB();

  const parsedDocuments = await MultiDocumentParser.parseDocuments(
    rawDocuments,
    {
      metadataDB,
      // Always allow by default. Use update_access_control script to change
      accessControlPolicyFactory: new AlwaysAllowDocumentAccessPolicyFactory(),
    }
  );

  const documentTransformer = new SeparatorTextChunker({ metadataDB });
  const transformedDocuments =
    await documentTransformer.transformDocuments(parsedDocuments);

  // Persist metadataDB to disk for loading in the other scripts
  await metadataDB.persist(
    "examples/typescript/financial_report/metadataDB.json"
  );

  // Use a new namespace for each run so that we can easily change which data to use
  const namespace = uuid();
  console.log(`NAMESPACE: ${namespace}`);

  await PineconeVectorDB.fromDocuments(transformedDocuments, {
    // Make sure this matches your Pinecone index name & it has 1536 dimensions for openai embeddings
    indexName: "test-financial-report",
    namespace,
    embeddings: new OpenAIEmbeddings(),
    metadataDB,
  });

  console.log("Ingestion complete");
}

main();

export {};
